File: 15/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        set<vector<int>> temp;
        sort(nums.begin(), nums.end());
        
        for(int i=0; i<nums.size(); ++i){
            if(nums[i] > 0)break;
            if(i>0 && nums[i] == nums[i-1])continue;
            int target = -nums[i];
            unordered_map<int, int> map;
            for(int j=i+1; j<nums.size(); j++){
                if(map.find(target-nums[j]) == map.end())
                    map[nums[j]] = j;
                else{
                    vector<int> vec{nums[i], nums[map[target-nums[j]]], nums[j]};
                    temp.insert(vec);
                }
            }
        }
        for(auto a:temp)
            ans.push_back(a);
        return ans;
    }
};

File: 15/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 3:
            if sum(nums) == 0:
                return [nums]
            else:
                return []

        nums.sort()
        ans = []
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j = i+1
            k = len(nums)-1
            while j<k:
                temp = nums[i]+nums[j]+nums[k]
                if temp > 0:
                    k -= 1
                elif temp < 0:
                    j += 1
                else:
                    ans.append([nums[i], nums[j], nums[k]])
                    j += 1
                    while nums[j] == nums[j-1] and j < k:
                        j += 1

        return ans
